### PR TITLE
Reader refactor

### LINKED ISSUE #228: Sample sheet parsing
In Seqlab 2, some of the sample sheet columns have changed:

- Sample_ID and Sample_Name have swapped roles
- Index and Index2 are no longer capitalised
- Project_Name no longer needs to be converted to Sample_Project

Some of the fuzzy logic for locating columns can probably be removed as well, since our sample sheet format is fairly stable now.

### PR DESCRIPTION
**Pull request by mwhamgenomics**
This accommodates the Seqlab 2 sample sheet changes described in #228. The design of SampleSheet and RunInfo has also been refactored.

Closes #228.

### CODE REVIEW

File: analysis_driver/report_generation/report_crawlers.py [hunk 1]
```diff
@@ -74,38 +74,33 @@ def _update_doc_set(d, k, v):
...
+        """
         self.barcodes_info = defaultdict(dict)
         self.unexpected_barcodes = {}
         self.libraries = defaultdict(dict)
         self.lanes = defaultdict(dict)
         self.run = {ELEMENT_RUN_NAME: self.run_id, ELEMENT_RUN_ELEMENTS: []}
         self.projects = defaultdict(dict)
 
-        for project_id, proj_obj in samplesheet.sample_projects.items():
+        for project_id, proj_obj in samplesheet.projects.items():
```
Comment (tcezard) [inline analysis_driver/report_generation/report_crawlers.py]: We should use `samplesheet.projects.values()` here and get the project id from `proj_obj.name`

File: analysis_driver/report_generation/report_crawlers.py [hunk 2]
```diff
@@ -114,7 +109,7 @@ def _populate_barcode_info_from_sample_sheet(self, samplesheet):
                         lib[ELEMENT_LIBRARY_INTERNAL_ID] = sample.sample_name
                         self._update_doc_list(lib, k=ELEMENT_RUN_ELEMENTS, v=run_element_id)
 
-                        # Populate the projects
+                        # Populate the sample_projects
```
Comment (tcezard) [inline analysis_driver/report_generation/report_crawlers.py]: I don't think this comment makes sense

File: analysis_driver/report_generation/report_crawlers.py [hunk 3]
```diff
@@ -293,7 +287,7 @@ def send_data(self):
                     pp('lanes', self.lanes.values(), ELEMENT_LANE_ID),
                     pp('runs', [self.run], ELEMENT_RUN_NAME),
                     pp('samples', self.libraries.values(), ELEMENT_SAMPLE_INTERNAL_ID, ['run_elements']),
-                    pp('projects', self.projects.values(), ELEMENT_PROJECT_ID, ['samples'])
+                    pp('sample_projects', self.projects.values(), ELEMENT_PROJECT_ID, ['samples'])
```
Comment (tcezard) [inline analysis_driver/report_generation/report_crawlers.py]: The name of the endpoint is `projects` not `sample_projects`

**Review by tcezard — CHANGES_REQUESTED** [review]

**Comment by tcezard** [comment]
Could you also add a changelog entry?

File: analysis_driver/report_generation/report_crawlers.py (hunk 3 above)
Comment (mwhamgenomics) [reply analysis_driver/report_generation/report_crawlers.py]: Silly IDE refactoring...

**Review by tcezard — APPROVED** [review]

**Pull request merged**